### DISCUSSION TITLE
Welcome to SCXML-tutorial Discussions!

### DISCUSSION THREAD

**Discussion by alexzhornyak** [alexzhornyak/SCXML-tutorial#6] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

**Comment by gaillardou31** [comment]
Hi, thanks for your tutorial.
Have you any example for including a datamodel from an XML file ?

> What is the format of the XML file ?
> What is the way to call the parameters from the XML file in the SCXML ?

I want to declare parameters in an XML file and use them in the SCXML.

BR,
Nicolas

**Reply by alexzhornyak** [reply]
Hi, there is a standard example [test557.scxml](https://github.com/alexzhornyak/SCXML-tutorial/blob/master/Tests/ecma/W3C/Optional/Auto/test557.scxml)

![image](https://user-images.githubusercontent.com/18611095/137474306-3f09621d-0f45-4353-b93f-1b672ecb6a96.png)

Block of XML is loaded as ecmascript object and then you should use domparser functions to access its data.

P.S. Be careful, support of XML in datamodel is optional and not all platforms support it (for example: Qt does not support).

**Reply by gaillardou31** [reply]
Thanks a lot, I am using Qt. I will try with JSON and Qt.

**Reply by gaillardou31** [reply]
As written in this Compliance web page (https://doc.qt.io/qt-5/qtscxml-scxml-compliance.html) JSON in not supported.
Have you any idea on how to declare typed data in an external file ?

**Reply by alexzhornyak** [reply]
![image](https://user-images.githubusercontent.com/18611095/137772203-cd7d381f-3b7f-4564-b5ff-c67dc14df85f.png)

**JSON is supported in Qt**. Here you are a simple example. You may use [ScxmlEditor](https://github.com/alexzhornyak/ScxmlEditor-Tutorial)- there is a builtin Qt Scxml Tester, so you may immediately check how Qt will behave itself with scxml

![image](https://user-images.githubusercontent.com/18611095/137772314-7bc2544a-2c0c-48d6-8bbd-36bbb5dbe576.png)
```xml
<scxml datamodel="ecmascript" name="ScxmlShape1" version="1.0" xmlns="http://www.w3.org/2005/07/scxml">
	<datamodel>
		<data id="VarJSON">{
  'array': [
    1,
    2,
    3
  ],
  'boolean': true,
  'color': 'gold',
  'null': null,
  'number': 123,
  'object': {
    'a': 'b',
    'c': 'd'
  },
  'string': 'Hello World'
}
		</data>
	</datamodel>
	<state id="testJSON1">
		<onentry>
			<log expr="VarJSON.string" label="test"/>
		</onentry>
		<transition cond="VarJSON.string == 'Hello World'" target="testJSON2"/>
	</state>
	<state id="testJSON2">
		<transition cond="VarJSON.boolean == true" target="testJSON3"/>
	</state>
	<final id="testJSON3"/>
</scxml>
```

**Reply by gaillardou31** [reply]
It works !
And if I package the JSON elements in a JSON file, and call them with a src="", it still works.

Thanks a lot for your help.

**Comment by heuripedes** [comment]
Does SCXML support AND, OR, XOR etc in the boolean expressions? Is it supported in the Editor too?

I've tried `In(S1) && In(S2)` and also `In(S1) and In(S2)` but it doesn't work. Using just `In(S1)` or `In(S2)` works.

*EDIT*:

> 5.9.1 Conditional Expressions: [...] The set of operators in conditional expressions varies depending on the data model [...]

Setting my scxml document's datamodel to ecmascript and quoting the state names worked. Sorry for bothering.

**Reply by alexzhornyak** [reply]
Hi, if you are using ScxmlEditor, there are a couple of capabilities to check chart:
**1) BETA: experimental validation of 'In' predicate (works for simple checks)**

![image](https://user-images.githubusercontent.com/18611095/141829774-c3700066-a677-4aff-a4ca-c4f67f768757.png)
![image](https://user-images.githubusercontent.com/18611095/141830397-2a4d381b-a819-40b8-9065-a4a1ad86d87d.png)

**2) expression syntax check (available in syntax editor of condition)**

![image](https://user-images.githubusercontent.com/18611095/141830847-1dd3ddb0-0400-48d4-a90f-b172b139bb94.png)

P.S. If you don't see 'EcmaScript' word in syntax editor, then datamodel is NULL
